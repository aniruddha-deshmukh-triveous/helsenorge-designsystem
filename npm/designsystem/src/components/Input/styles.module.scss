@import '../../scss/_spacers.scss';
@import '../../scss/_breakpoints.scss';
@import '../../scss/_palette.scss';
@import '../../scss/_font-settings.scss';
@import '../../scss/_input.scss';

.input-wrapper {
  @include input-wrapper;

  margin: getSpacer(l) 0;
}

.input-wrapper:first-of-type {
  margin-top: 0;
}

.input-wrapper:last-of-type {
  margin-bottom: 0;
}

.content-wrapper {
  @include content-wrapper;

  padding: 0.5rem 0;
  display: inline-flex;
  align-items: center;

  &--with-icon {
    padding: 1px 0;

    @media (min-width: map-get($grid-breakpoints, md)) {
      padding: 0.375rem 0;
    }
  }

  &__input {
    line-height: getSpacer(m);
    height: getSpacer(m);

    &--bigform {
      height: 2.5rem;
    }

    @media (min-width: map-get($grid-breakpoints, md)) {
      line-height: 1.625rem;
      height: 2.125rem;
      font-size: $font-size-md;

      &--bigform {
        line-height: 1.75rem;
        height: 3.5rem;
      }
    }
  }
}
