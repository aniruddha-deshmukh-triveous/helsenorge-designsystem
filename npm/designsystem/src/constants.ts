export enum IconSize {
  XSmall = 38,
  Small = 48,
  Medium = 64,
  Large = 80,
  XLarge = 130,
}

export enum ZIndex {
  Modal = 1300000,
}

export interface HTMLButtonProps {
  autoFocus?: boolean;
  disabled?: boolean;
  form?: string;
  formAction?: string;
  formEncType?: string;
  formMethod?: string;
  formNoValidate?: boolean;
  formTarget?: string;
  name?: string;
  type?: 'submit' | 'reset' | 'button' | string;
  value?: string | string[] | number;
}

export interface HTMLTextareaProps {
  autoFocus?: boolean;
  disabled?: boolean;
  form?: string;
  formNoValidate?: boolean;
  formTarget?: string;
  name?: string;
  id?: string;
  autocomplete?: 'off' | 'on';
  placeholder?: string;
  readonly?: boolean;
  required?: boolean;
  spellcheck?: 'true' | 'default' | 'false';
  'aria-invalid'?: boolean;
}

export interface HTMLAnchorProps {
  download?: string;
  href?: string;
  hrefLang?: string;
  media?: string;
  ping?: string;
  rel?: string;
  target?: string;
  referrerPolicy?:
    | ''
    | 'same-origin'
    | 'no-referrer'
    | 'no-referrer-when-downgrade'
    | 'origin'
    | 'origin-when-cross-origin'
    | 'strict-origin'
    | 'strict-origin-when-cross-origin'
    | 'unsafe-url';
}

export type ButtonVariant = 'secondary' | 'tertiary' | string | undefined | null;

export enum FormMode {
  onwhite = 'onwhite',
  ongrey = 'ongrey',
  onblueberry = 'onblueberry',
  ondark = 'ondark',
  oninvalid = 'oninvalid',
}

export enum FormVariant {
  normal = 'normal',
  bigform = 'bigform',
}
