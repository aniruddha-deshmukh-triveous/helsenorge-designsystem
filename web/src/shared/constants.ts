export const HNDS_BUCKET_IMAGE_URL = ''; // Kan endres her hvis bilder skal hostes et annet sted

export interface StyledDefaultProps {
  children?: React.ReactNode;
  className?: string;
}
