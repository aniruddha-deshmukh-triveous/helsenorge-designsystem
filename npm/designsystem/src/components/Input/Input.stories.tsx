import React from 'react';
import { withKnobs, text, select, boolean } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import Input, { InputTypes } from './Input';
import { withA11y } from '@storybook/addon-a11y';
import { mapToBackgoundColor } from '../../../.storybook/StoryBackground';

import { FormMode, FormVariant } from '../../constants';
import Hospital from '../Icons/Hospital';
import Icon, { IconSize } from '../Icons';

const stories = storiesOf('Input', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => {
  const modes = select('background', FormMode, FormMode.onwhite);

  return (
    <div style={{ width: '80vw', background: mapToBackgoundColor(modes), padding: 20 }}>
      <Input
        defaultValue={text('defaultValue', '')}
        placeholder={text('placeholdertext', 'Skriv inn tekst her')}
        testId={'123-test'}
        variant={select('Variant', FormVariant, FormVariant.normal)}
        transparent={boolean('transparent', false)}
        type={select('type', InputTypes, InputTypes.text)}
        disabled={boolean('disabled', false)}
        mode={modes}
        label={'Skriv inn din tekst'}
        icon={boolean('icon', true) ? Hospital : undefined}
        iconRight={boolean('iconRight', false)}
        readOnly={boolean('readOnly', false)}
      />
    </div>
  );
});

stories.add('Multiple examples', () => {
  const modes = select('background', FormMode, FormMode.onwhite);

  return (
    <div style={{ width: '80vw', background: mapToBackgoundColor(modes), padding: 20 }}>
      <Input
        defaultValue={text('defaultValue', '')}
        placeholder={text('placeholdertext', 'Skriv inn tekst her')}
        testId={'123-test'}
        variant={select('Variant', FormVariant, FormVariant.normal)}
        transparent={boolean('transparent', false)}
        type={select('type', InputTypes, InputTypes.text)}
        disabled={boolean('disabled', false)}
        mode={modes}
        label={'Skriv inn din tekst'}
        icon={boolean('icon', true) ? Hospital : undefined}
        iconRight={boolean('iconRight', false)}
        readOnly={boolean('readOnly', false)}
      />

      <Input
        defaultValue={text('defaultValue', '')}
        placeholder={text('placeholdertext', 'Skriv inn tekst her')}
        testId={'123-test'}
        variant={select('Variant', FormVariant, FormVariant.normal)}
        transparent={boolean('transparent', false)}
        type={select('type', InputTypes, InputTypes.text)}
        disabled={boolean('disabled', false)}
        mode={modes}
        label={'Skriv inn din tekst'}
        icon={boolean('icon', true) ? Hospital : undefined}
        iconRight={boolean('iconRight', false)}
        readOnly={boolean('readOnly', false)}
      />

      <Input
        defaultValue={text('defaultValue', '')}
        placeholder={text('placeholdertext', 'Skriv inn tekst her')}
        testId={'123-test'}
        variant={select('Variant', FormVariant, FormVariant.normal)}
        transparent={boolean('transparent', false)}
        type={select('type', InputTypes, InputTypes.text)}
        disabled={boolean('disabled', false)}
        mode={modes}
        label={'Skriv inn din tekst'}
        icon={boolean('icon', true) ? Hospital : undefined}
        iconRight={boolean('iconRight', false)}
        readOnly={boolean('readOnly', false)}
      />
    </div>
  );
});

stories.add('Children after label', () => {
  const modes = select('background', FormMode, FormMode.onwhite);

  return (
    <div style={{ width: '80vw', background: mapToBackgoundColor(modes), padding: 20 }}>
      <Input
        defaultValue={text('defaultValue', '')}
        placeholder={text('placeholdertext', 'Skriv inn tekst her')}
        testId={'123-test'}
        variant={select('Variant', FormVariant, FormVariant.normal)}
        transparent={boolean('transparent', false)}
        type={select('type', InputTypes, InputTypes.text)}
        disabled={boolean('disabled', false)}
        mode={modes}
        label={'Skriv inn din tekst'}
        icon={boolean('icon', true) ? Hospital : undefined}
        iconRight={boolean('iconRight', false)}
        readOnly={boolean('readOnly', false)}
        afterLabelChildren={<Icon size={IconSize.XSmall} svgIcon={Hospital}></Icon>}
      />
    </div>
  );
});
