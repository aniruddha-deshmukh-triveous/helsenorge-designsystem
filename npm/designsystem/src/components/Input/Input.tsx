import React, { useRef } from 'react';
import cn from 'classnames';

import { FormMode, FormVariant, AnalyticsId } from '../../constants';
import { uuid } from '../../utils/uuid';
import Icon, { IconSize, SvgIcon } from '../Icons';
import { getColor } from '../../theme/currys';
import { Breakpoint, useBreakpoint } from '../../hooks/useBreakpoint';

import styles from './styles.module.scss';

export interface InputProps {
  /** initial value for input */
  defaultValue?: string;
  /** Text for the placeholder */
  placeholder?: string;
  /**  HMTL Input type */
  type?: keyof typeof InputTypes;
  /** input id */
  inputId?: string;
  /** Unique identifyer for the input tag */
  name?: string;
  /** If true, the component will be transparent. */
  transparent?: boolean;
  /** Icon to be displayed next to the input field */
  icon?: SvgIcon;
  /** Places the icon to the right */
  iconRight?: boolean;
  /** Changes the color profile of the input */
  mode?: keyof typeof FormMode;
  /** Changes the visuals of the input */
  variant?: keyof typeof FormVariant;
  /** Label of the input */
  label?: string;
  /** Activates Error style for the input */
  error?: boolean;
  /** Error text to show above the component */
  errorText?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Disables the input field */
  disabled?: boolean;
  /** Om input feltet er readOnly */
  readOnly?: boolean;
  /** attribute autocomplete,  */
  autocomplete?: string;
  /** Component shown after label */
  afterLabelChildren?: React.ReactNode;
  /** Component shown under label */
  belowLabelChildren?: React.ReactNode;
}

export enum InputTypes {
  text = 'text',
  number = 'number',
  email = 'email',
  password = 'password',
  search = 'search',
  tel = 'tel',
  url = 'url',
}

const Input = React.forwardRef((props: InputProps, ref: React.Ref<HTMLInputElement>) => {
  const {
    defaultValue,
    placeholder,
    type = InputTypes.text,
    inputId = uuid(),
    name,
    transparent = false,
    icon,
    iconRight,
    mode,
    variant,
    label,
    error,
    errorText,
    testId,
    disabled,
    readOnly,
    autocomplete,
    afterLabelChildren,
    belowLabelChildren,
    ...restProps
  } = props;
  const breakpoint = useBreakpoint();
  const contentWrapperRef = useRef<HTMLDivElement>(null);

  const onDark = mode === FormMode.ondark;
  const onBlueberry = mode === FormMode.onblueberry;
  const onError = mode === FormMode.oninvalid || errorText || error;
  const bigForm = variant === FormVariant.bigform;
  const isTransparent = transparent && mode !== FormMode.ondark && !onError;

  const inputWrapperClass = cn(styles['input-wrapper'], {
    [styles['input-wrapper--invalid']]: errorText,
  });

  const labelWrapperClass = cn(styles['input-wrapper__label-wrapper'], {
    [styles['input-wrapper__label-wrapper--on-dark']]: onDark,
    [styles['input-wrapper__label-wrapper--disabled']]: disabled,
  });

  const contentWrapperClass = cn(styles['content-wrapper'], {
    [styles['content-wrapper--transparent']]: isTransparent,
    [styles['content-wrapper--on-blueberry']]: onBlueberry,
    [styles['content-wrapper--on-dark']]: onDark,
    [styles['content-wrapper--invalid']]: onError,
    [styles['content-wrapper--bigform']]: bigForm,
    [styles['content-wrapper--disabled']]: disabled,
    [styles['content-wrapper--with-icon']]: icon,
  });

  const inputClass = cn(styles['content-wrapper__input'], {
    [styles['content-wrapper__input--bigform']]: bigForm,
    [styles['content-wrapper__input--disabled']]: disabled,
  });

  const iconColor = disabled ? getColor('neutral', 500) : getColor('black');

  const renderIcon = () => {
    const iconSize = breakpoint === Breakpoint.xs || !bigForm ? IconSize.XSmall : IconSize.Small;
    return icon !== undefined ? <Icon color={iconColor} size={iconSize} svgIcon={icon} /> : null;
  };

  const handleClick = (): void => {
    if (contentWrapperRef && contentWrapperRef.current && icon) {
      const selectedChild = iconRight ? 0 : 1;
      const input = contentWrapperRef.current.children[selectedChild] as HTMLInputElement;
      input.focus();
    }
  };

  return (
    <div data-testid={testId} data-analyticsid={AnalyticsId.Input} className={inputWrapperClass}>
      {label && (
        <div className={labelWrapperClass}>
          <label htmlFor={inputId}>{label}</label>
          {afterLabelChildren && <div className={styles['input-wrapper__after-label-children']}>{afterLabelChildren}</div>}
        </div>
      )}
      {belowLabelChildren && <div>{belowLabelChildren}</div>}
      <div onClick={handleClick} ref={contentWrapperRef} className={contentWrapperClass}>
        {!iconRight && renderIcon()}
        <input
          name={name}
          type={type}
          defaultValue={defaultValue}
          id={inputId}
          className={inputClass}
          ref={ref}
          aria-invalid={!!onError}
          disabled={disabled}
          placeholder={placeholder}
          readOnly={readOnly}
          autoComplete={autocomplete || 'off'}
          {...restProps}
        />
        {iconRight && renderIcon()}
      </div>
    </div>
  );
});

export default Input;
