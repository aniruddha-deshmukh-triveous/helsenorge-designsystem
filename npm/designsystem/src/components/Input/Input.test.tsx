import React from 'react';
import { render, screen } from '@testing-library/react';
import Input, { InputTypes } from './Input';
import { FormMode, FormVariant } from '../../constants';
import Hospital from '../Icons/Hospital';

describe('Gitt at Input skal vises', (): void => {
  describe('Når Input rendres', (): void => {
    test('Så vises Input', (): void => {
      const { container } = render(<Input label={'Skriv noe'} />);

      expect(container).toMatchSnapshot();

      const label = screen.getByText('Skriv noe');
      expect(label).toBeVisible();

      const input = screen.getByRole('textbox');
      expect(input).toBeVisible();
      expect(input).toHaveClass('content-wrapper__input');
    });
  });

  describe('Når disabled er true', (): void => {
    test('Så vises Input som disabled', (): void => {
      render(<Input label={'Skriv noe!'} disabled />);

      const input = screen.getByRole('textbox');
      expect(input).toBeDisabled();
    });
  });

  describe('Når FormMode er onblueberry', (): void => {
    test('Så vises Input med onblueberry styling', (): void => {
      render(<Input label={'Skriv noe'} mode={FormMode.onblueberry} />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--on-blueberry');
    });
  });
  describe('Når mode er ondark', (): void => {
    test('Så vises Input med ondark styling', (): void => {
      render(<Input label={'Skriv noe'} mode={FormMode.ondark} />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--on-dark');
    });
  });
  describe('Når transparent er true', (): void => {
    test('Så vises Input med transparent styling', (): void => {
      render(<Input label={'Skriv noe'} transparent />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--transparent');
    });
  });
  describe('Når formvariant er bigform', (): void => {
    test('Så vises Input med bigform styling', (): void => {
      render(<Input label={'Skriv noe'} variant={FormVariant.bigform} />);

      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--bigform');
    });
  });

  describe('Når Input har en default value', (): void => {
    test('Så skal default value vises', (): void => {
      render(<Input label={'Skriv noe'} defaultValue={'default text'} />);
      const input = screen.getByRole('textbox');
      expect(input).toHaveValue('default text');
    });
  });

  describe('Når Input har en placeholder value', (): void => {
    test('Så skal placeholder verdien vises', (): void => {
      render(<Input label={'Skriv noe'} placeholder={'placeholder  text'} />);
      const input = screen.getByPlaceholderText('placeholder text');
      expect(input).toBeVisible();
    });
  });

  describe('Når type settes', (): void => {
    test('Så skal input ha den typen satt', (): void => {
      render(<Input label={'Skriv noe'} type={InputTypes.number} />);
      const input = screen.getByRole('spinbutton');
      expect(input).toBeVisible();
    });
  });

  describe('Når icon settes', (): void => {
    test('Så rendres input med icon', (): void => {
      render(<Input label={'Skriv noe'} icon={Hospital} />);
      const svg = screen.getByRole('textbox').previousSibling;
      expect(svg).toHaveClass('hnds-style-icon');
    });
  });

  describe('Når iconRight er true', (): void => {
    test('Så rendres input med icon til høyre', (): void => {
      render(<Input label={'Skriv noe'} icon={Hospital} iconRight />);
      const svg = screen.getByRole('textbox').nextSibling;
      expect(svg).toHaveClass('hnds-style-icon');
    });
  });

  describe('Når Input har en error', (): void => {
    test('Så skal invalid styling brukes', (): void => {
      render(<Input label={'Skriv noe'} error />);
      const contentWrapper = screen.getByRole('textbox').parentElement;
      expect(contentWrapper).toHaveClass('content-wrapper content-wrapper--invalid');
    });
  });
});
