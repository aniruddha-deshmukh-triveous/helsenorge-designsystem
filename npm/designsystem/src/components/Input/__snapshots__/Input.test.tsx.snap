// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Input skal vises Når Input rendres Så vises Input 1`] = `
<div>
  <div
    class="input-wrapper"
    data-analyticsid="input"
  >
    <div
      class="input-wrapper__label-wrapper"
    >
      <label
        for="-unik-id"
      >
        Skriv noe
      </label>
    </div>
    <div
      class="content-wrapper"
    >
      <input
        aria-invalid="false"
        autocomplete="off"
        class="content-wrapper__input"
        id="-unik-id"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
